File: spk/pysolve/legacy/_solver.py
# Copyright (c) 2021 Sony Pictures Imageworks, et al.
# SPDX-License-Identifier: Apache-2.0
# https://github.com/imageworks/spk

from typing import List, Union, Dict

from ruamel import yaml
import structlog

from ... import api, storage
from .. import Solution
from ._package_iterator import (
    RepositoryPackageIterator,
    FilteredPackageIterator,
)
from ._decision import Decision, DecisionTree
from ._errors import SolverError, UnresolvedPackageError, ConflictingRequestsError

_LOGGER = structlog.get_logger("spk.solve.legacy")


class Solver:
    """Solver is the main entrypoint for resolving a set of packages."""

    def __init__(self, options: Union[api.OptionMap, Dict[str, str]]) -> None:

        self._repos: List[storage.Repository] = []
        self.decision_tree = DecisionTree()
        self.decision_tree.root.update_options(options)
        self._running = False
        self._complete = False
        self._binary_only = False

    def reset(self) -> None:

        self._repos = []
        self.decision_tree = DecisionTree()
        self._running = False
        self._complete = False
        self._binary_only = False

    def set_binary_only(self, binary_only: bool) -> None:
        """If true, only solve pre-built binary packages.

        When false, the solver may return packages where the build is not set.
        These packages are known to have a source package available, and the requested
        options are valid for a new build of that source package.
        These packages are not actually built as part of the solver process but their
        build environments are fully resolved and dependencies included
        """
        self._binary_only = binary_only

    def update_options(self, options: Union[Dict[str, str], api.OptionMap]) -> None:

        self.decision_tree.root.update_options(options)

    def add_repository(self, repo: storage.Repository) -> None:
        """Add a repository where the solver can get packages."""

        self._repos.append(repo)

    def add_request(self, pkg: Union[str, api.Ident, api.Request]) -> None:
        """Add a package request to this solver.

        Raises:
            RuntimeError: if the solver has already completed
        """

        if self._complete:
            raise RuntimeError("Solver has already been executed")
        self.decision_tree.root.add_request(pkg)

    def solve_build_environment(self, spec: api.Spec) -> Solution:
        """Adds requests for all build """

        build_options = spec.resolve_all_options(self.decision_tree.root.get_options())
        for option in spec.build.options:
            if not isinstance(option, api.PkgOpt):
                continue
            given = build_options.get(option.pkg)
            request = option.to_request(given)
            self.add_request(request)

        return self.solve()

    def solve(self) -> Solution:
        """Solve the current set of package requests into a complete environment.

        Raises:
            RuntimeError: if the solver has already completed
        """

        if self._complete:
            raise RuntimeError("Solver has already been executed")

        self._running = True

        state = self.decision_tree.root
        request = state.next_request()
        while request is not None:

            if request.pin:
                _LOGGER.warning(
                    "Solving for unpinned request, this is probably not what you want to be happening!",
                    request=request,
                )

            try:
                state = self._solve_request(state, request)
            except SolverError:
                if state.parent is None:
                    stack = self.decision_tree.get_error_chain()
                    raise stack[-1] from None
                state = state.parent

            request = state.next_request()

        self._running = False
        self._complete = True
        solution = state.get_current_solution()
        return solution

    def _solve_request(self, state: Decision, request: api.PkgRequest) -> Decision:

        decision = state.add_branch()
        iterator = state.get_iterator(request.pkg.name)
        if iterator is None:
            iterator = self._make_iterator(decision, request)
            state.set_iterator(request.pkg.name, iterator)

        try:

            while True:
                spec, repo = next(iterator)
                if spec.pkg.build is None:
                    if self._binary_only:
                        compat = api.Compatibility("Only binary packages are allowed")
                    else:
                        compat = self._resolve_new_build(spec, state)
                    if not compat:
                        iterator.add_history(spec.pkg, compat)
                        continue
                elif not spec.pkg.build == api.SRC:
                    try:
                        with decision.transaction() as t:
                            for dep in spec.install.requirements:
                                t.add_request(dep)
                    except ConflictingRequestsError as e:
                        iterator.add_history(spec.pkg, api.Compatibility(str(e)))
                        continue
                break

            decision.set_resolved(spec, repo)

        except StopIteration:
            err = UnresolvedPackageError(
                yaml.safe_dump(request.to_dict()).strip(),  # type: ignore
                history=iterator.get_history(),
            )
            decision.set_error(err)
            raise err from None
        except SolverError as e:
            decision.set_error(e)
            raise

        return decision

    def _make_iterator(
        self, state: Decision, request: api.PkgRequest
    ) -> FilteredPackageIterator:

        assert len(self._repos), "No configured package repositories."
        return FilteredPackageIterator(
            RepositoryPackageIterator(request.pkg.name, self._repos),
            request,
            state.get_options(),
        )

    def _resolve_new_build(self, spec: api.Spec, state: Decision) -> api.Compatibility:

        opts = state.get_options()
        solver = Solver(opts)
        for repo in self._repos:
            solver.add_repository(repo)

        try:
            solution = solver.solve_build_environment(spec)
        except SolverError as err:
            return api.Compatibility(f"Failed to resolve build env: {err}")

        spec = spec.copy()
        spec.update_spec_for_build(opts, list(s for _, s, _ in solution.items()))
        for request in spec.install.requirements:
            try:
                state.add_request(request)
            except ConflictingRequestsError as err:
                return api.Compatibility(str(err))

        return api.COMPATIBLE

File: crates/spk-solve/src/solver.rs
// Copyright (c) Contributors to the SPK project.
// SPDX-License-Identifier: Apache-2.0
// https://github.com/spkenv/spk

use std::borrow::Cow;
use std::sync::Arc;

use enum_dispatch::enum_dispatch;
use spk_schema::ident::{PinnedValue, PkgRequestWithOptions, RequestWithOptions, VarRequest};
use spk_schema::{OptionMap, Recipe};
use spk_solve_solution::Solution;
use spk_storage::RepositoryHandle;
use variantly::Variantly;

use crate::{DecisionFormatter, Result};

#[enum_dispatch(Solver, SolverExt, SolverMut)]
// Don't derive Default. If some code is generic on Solver and is given one of
// these, if it wants a "default" solver it needs to be given a new solver of
// the same variety and `SolverImpl::default()` can't do that.
#[derive(Clone, Variantly)]
pub enum SolverImpl {
    Step(crate::StepSolver),
    Resolvo(crate::solvers::ResolvoSolver),
}

#[async_trait::async_trait]
#[enum_dispatch]
pub trait Solver {
    /// Return the options that the solver is currently configured with.
    ///
    /// These are the options that have been set via
    /// [`SolverMut::update_options`].
    fn get_options(&self) -> Cow<'_, OptionMap>;

    /// Return the PkgRequests added to the solver.
    fn get_pkg_requests(&self) -> Vec<PkgRequestWithOptions>;

    /// Return the VarRequests added to the solver.
    fn get_var_requests(&self) -> Vec<VarRequest<PinnedValue>>;

    /// Return a reference to the solver's list of repositories.
    fn repositories(&self) -> &[Arc<RepositoryHandle>];
}

#[async_trait::async_trait]
#[enum_dispatch]
pub trait SolverMut: Solver {
    /// Add a request to this solver.
    fn add_request(&mut self, request: RequestWithOptions);

    /// Adds requests for all build requirements of the given recipe.
    fn configure_for_build_environment<T: Recipe>(&mut self, recipe: &T) -> Result<()> {
        let options = self.get_options();

        let build_options = recipe.resolve_options(&*options)?;
        for request in recipe
            .get_build_requirements(&build_options)?
            .iter()
            .cloned()
        {
            self.add_request(request)
        }

        Ok(())
    }

    /// Put this solver back into its default state
    fn reset(&mut self);

    /// Run the solver as configured using the given formatter.
    ///
    /// "log" means that solver progress is output via tracing, as
    /// configured by the formatter.
    ///
    /// The solution may also be printed, if found, as configured by the
    /// formatter.
    ///
    /// Not all formatter settings may be supported by every solver.
    async fn run_and_log_resolve(&mut self, formatter: &DecisionFormatter) -> Result<Solution>;

    /// Run the solver as configured using the given formatter.
    ///
    /// "print" means that solver progress is printed to the console, as
    /// configured by the formatter.
    ///
    /// The solution may also be printed, if found, as configured by the
    /// formatter.
    ///
    /// Not all formatter settings may be supported by every solver.
    async fn run_and_print_resolve(&mut self, formatter: &DecisionFormatter) -> Result<Solution>;

    /// If true, only solve pre-built binary packages.
    ///
    /// When false, the solver may return packages where the build is not set.
    /// These packages are known to have a source package available, and the requested
    /// options are valid for a new build of that source package.
    /// These packages are not actually built as part of the solver process but their
    /// build environments are fully resolved and dependencies included
    fn set_binary_only(&mut self, binary_only: bool);

    /// Run the solver as configured.
    async fn solve(&mut self) -> Result<Solution>;

    fn update_options(&mut self, options: OptionMap);
}

impl<T> Solver for &T
where
    T: Solver,
{
    fn get_options(&self) -> Cow<'_, OptionMap> {
        T::get_options(self)
    }

    fn get_pkg_requests(&self) -> Vec<PkgRequestWithOptions> {
        T::get_pkg_requests(self)
    }

    fn get_var_requests(&self) -> Vec<VarRequest<PinnedValue>> {
        T::get_var_requests(self)
    }

    fn repositories(&self) -> &[Arc<RepositoryHandle>] {
        T::repositories(self)
    }
}

impl<T> Solver for &mut T
where
    T: Solver,
{
    fn get_options(&self) -> Cow<'_, OptionMap> {
        T::get_options(self)
    }

    fn get_pkg_requests(&self) -> Vec<PkgRequestWithOptions> {
        T::get_pkg_requests(self)
    }

    fn get_var_requests(&self) -> Vec<VarRequest<PinnedValue>> {
        T::get_var_requests(self)
    }

    fn repositories(&self) -> &[Arc<RepositoryHandle>] {
        T::repositories(self)
    }
}

#[async_trait::async_trait]
impl<T> SolverMut for &mut T
where
    T: SolverMut + Send + Sync,
{
    fn add_request(&mut self, request: RequestWithOptions) {
        T::add_request(self, request)
    }

    fn reset(&mut self) {
        T::reset(self)
    }

    async fn run_and_log_resolve(&mut self, formatter: &DecisionFormatter) -> Result<Solution> {
        T::run_and_log_resolve(self, formatter).await
    }

    async fn run_and_print_resolve(&mut self, formatter: &DecisionFormatter) -> Result<Solution> {
        T::run_and_print_resolve(self, formatter).await
    }

    fn set_binary_only(&mut self, binary_only: bool) {
        T::set_binary_only(self, binary_only)
    }

    async fn solve(&mut self) -> Result<Solution> {
        T::solve(self).await
    }

    fn update_options(&mut self, options: OptionMap) {
        T::update_options(self, options)
    }
}

#[async_trait::async_trait]
#[enum_dispatch]
pub trait SolverExt: Solver {
    /// Add a repository where the solver can get packages.
    fn add_repository<R>(&mut self, repo: R)
    where
        R: Into<Arc<RepositoryHandle>>;
}

impl<T> SolverExt for &mut T
where
    T: SolverExt + Sync,
{
    fn add_repository<R>(&mut self, repo: R)
    where
        R: Into<Arc<RepositoryHandle>>,
    {
        T::add_repository(self, repo);
    }
}
